kernel/kernel-impl/src/main/java/org/sakaiproject/content/impl/BaseContentService.java

```diff
@@ -14374,26 +14374,45 @@ public void hardDelete(String siteId) {
 		}
 
 		// Cleanup the collections
-		String folderId = "";
-		try {
-			ContentCollection siteCollection = getCollection(collectionId);
-			List<ContentCollectionEdit> folders = m_storage.getCollections(siteCollection);
-			for (ContentCollectionEdit folder : folders) {
-				folderId = folder.getId();
-				log.debug("Removing collection: {}", folderId);
-				removeCollection(folder.getId());
-			}
-
-			// Now delete the site-root collection
-			log.debug("Removing collection: {}", siteCollection.getId());
-			removeCollection(siteCollection.getId());
-		} catch (IdUnusedException ide) {
-			log.warn("No resources in collection {}.", folderId);
-		} catch (Exception e) {
-			log.warn("Failed to remove collection {}.", folderId, e);
-		}
+		removeCollectionRecursive(collectionId);
     }
 
+	private void removeCollectionRecursive(String collectionId){
+		ContentCollection collection = null;
+		try{
+			collection = getCollection(collectionId);
+		}catch (IdUnusedException ide){
+			log.warn("No resources in collection {}.", collectionId);
+		}catch (TypeException te){
+			log.warn("Not a collection {}.", collectionId);
+		}catch (PermissionException pe){
+			log.warn("No permission to remove collection {}.", collectionId);
+		}
+		if(collection != null){
+			List<ContentCollectionEdit> members = m_storage.getCollections(collection);
+			if(members != null){
+				for(ContentCollectionEdit contentCollectionEdit:members){
+					removeCollectionRecursive(contentCollectionEdit.getId());
+				}
+			}
```

Comment:
I don't think getCollections ever returns null here. I've just checked the kernel code. So ...

```suggestion
			members.forEach(edit -> removeCollectionRecursive(edit.getId()));
```